rosbag2_compression/src/rosbag2_compression/compression_utils.hpp

```diff
@@ -0,0 +1,115 @@
+// Copyright 2020 Southwest Research Institute. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_COMPRESSION__COMPRESSION_UTILS_HPP_
+#define ROSBAG2_COMPRESSION__COMPRESSION_UTILS_HPP_
+
+#include <zstd.h>
+
+#include <chrono>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "logging.hpp"
+
+#include "rcpputils/filesystem_helper.hpp"
+
+namespace rosbag2_compression
+{
+// Increasing the compression level will:
+//   - Increase the time taken to compress
+//   - Decrease the size of the compressed data
+// Setting to zero uses Zstd's default value of 3.
+constexpr const int kDefaultZstdCompressionLevel = 1;
+// String constant used to identify ZstdCompressor.
+constexpr const char kCompressionIdentifier[] = "zstd";
+// String constant used to identify ZstdDecompressor.
+constexpr const char kDecompressionIdentifier[] = "zstd";
+// Used as a parameter type in a function that accepts the output of ZSTD_compress.
+using ZstdCompressReturnType = decltype(ZSTD_compress(
+    nullptr, 0,
+    nullptr, 0, 0));
+// Used as a parameter type in a function that accepts the output of ZSTD_decompress.
+using ZstdDecompressReturnType = decltype(ZSTD_decompress(
+    nullptr, 0,
+    nullptr, 0));
+// Used as a parameter type in a function that accepts the output of ZSTD_getFrameContentSize.
+using ZstdGetFrameContentSizeReturnType = decltype(ZSTD_getFrameContentSize(nullptr, 0));
+
+/**
+ * Open a file using the OS-specific C API.
+ * \param uri is the path to the file.
+ * \param read_mode is the read mode accepted by OS-specific fopen.
+ * \return the FILE pointer or nullptr if the file was not opened.
+ */
+FILE * open_file(const std::string & uri, const std::string & read_mode);
```

Comment:
Just want to double check, could we hide any of these functions private to the compression_utils compilation unit - or are they all being used elsewhere?

Reply:
I don't believe any of them are being used elsewhere, although I'm not familiar with what you mean by "hide any of these functions private to the compression_utils compilation unit"; is there an example of that somewhere?

Reply:
By that I mean don't declare them in the header, instead put them in an unnamed namespace in the cpp, e.g.

```
// header.hpp

void my_public_function();
```


```
// impl.cpp
#include "header.hpp"

namespace {
void some_helper_fn() {
}
}

void my_public_fn() {
  some_helper_fn();
}

```

Reply:
Oh, I see what you mean now.  I checked and `open_file` is actually the only one that's only being used in compression_utils, but I'll move it into the cpp file.